* TOC
{:toc}

## Biotech 7005: Frontiers of Biotechnology: Bioinformatics and Systems Modelling
{:.no_toc}

Semester 2 2017

Coordinator:
- Dan Kortschak

Instructors:
- Jimmy Breen
- Stephen Pederson
- Dan Kortschak
- Hien To

__*Course Contact*__
Dan Kortschak
Email: dan.kortschak@adelaide.edu.au
Phone: 08 8313 0928
Office: Braggs 2.60

### Contact Hours

Lectures: Thursday 9am, Barr Smith South 534 Forum Lecture Theatre
- 27 Jul — 14 Sep
- 5 Oct — 26 Oct

Tutorials: Friday 9am-11pm, Mawson, 134, Sprigg Room
- 28 Jul — 15 Sep
- 6 Oct — 27 Oct

Practicals: Monday 1pm-3pm, Mawson, G12 Computer Suite
- 24 Jul — 11 Sep
- 9 Oct — 23 Oct

### Timetable

| **Week** | **Monday** |**Lecture**    | **Practical/Tutorial** |
|----------|------------|---------------|---------------|
| **1** | 24/7 | Introduction to Bioinformatics (Dan)| [Introduction to R (Steve) (1)](Practicals/R_Practicals) |
| **2** | 31/7 | [Statistics (Steve)](Lectures/02-statistics.html) | [Introduction to R (Steve) (2)](Practicals/R_Practicals) |
| **3** | 7/8  | Evolutionary Processes (Dan) | [Introduction To Bash (Steve)](Practicals/Bash_Practicals/1_IntroBash.md) <br> [Tutorial: Statistics](Tutorials/Wk3_Statistics.html)|
| **4** | 14/8 | Alignment Process (Dan) | [Scripting In Bash (Steve)](Practicals/Bash_Practicals/2_BashScripting.md) |
| **5** | 21/8 | Alignment Process (Dan) | [Evolutionary processes (Dan) (1)](evolutionary_prac/evolutionary.md) |
| **6** | 28/8 | Next-generation sequencing (Dan) | NGS data practical (Steve/Jimmy) (1) |
| **7** | 4/9  | Genomics/Repeats/Metagenomics (Jimmy) | NGS data practical (Steve/Jimmy) (2) |
| **8** | 11/9 | Non-coding RNA/Transcriptomics (Steve) | [RNAseq practical (Steve)](Practicals/Transcriptome_Practical) |
| **-** | 18/9 | Semester Break |
| **-** | 25/9 | Semester Break |
| **9** | 2/10 (PH) | Epigenetics (Jimmy) | No practical (PH) |
| **10** | 9/10 | Systems Biology (Ville) | Systems Biology practical (Ville) (1) |
| **11** | 16/10 | Introduction to Systems Biology (Dan)  | [Graphical analyses (Dan) (1)](https://github.com/kortschak/graphprac)  |
| **12** | 23/10 | The Future — wooooooo (Dan) | Wrap-up (Jimmy/Dan/Steve)  |


## Assessment

### Assessment Schedule & Weighting

| **Assessment** | **Subject** | **Weight** | **Due Date** |
|----------------|-------------|------------|--------------|
| [Assessment 1](Assignments/Assignment1.md) | R programming | 7% | Friday 4th August |
| [Assessment 2](Assignments/Assignment2.md) | Bash Scripting | 7% | Monday 21st August |
| [Assessment 3](evolutionary_prac/evolutionary.md) | Evolutionary processes | 7% | Friday 8st September |
| [Assessment 4](Assignments/practical_assignment4_ngs.md) | NGS data practical | 7% | Friday 15th September |
| Assessment 5 | RNAseq practical | 7% | Friday 22th September |
| Assessment 6 | Graphical analyses | 7% | Monday 23rd October |
| Active Participation | | 3% | |
| Project | Complete Dataset | 15% | Friday 20th October |
| Exam | | 40% | |

### Assessment Checklist

Have you:

- [ ] Answered all the questions?
- [ ] Followed naming conventions for Assessments?
- [ ] Attached an assessment coversheet? (Available [here](http://www.sciences.adelaide.edu.au/current-students/forms/assessment%20cover%20sheet-28mar.pdf)).
	- This is a course requirement. See the [Academic Honesty Policy](http://www.adelaide.edu.au/policies/230/).
- [ ] Identified the work as yours?
	- Emails should have the course and assessment task names.
	- Documents should be named with your name, the course name and the assessment task.
	- Printed documents should have you name and the course and assessment task in the text/footer/header.
- [ ] Used appropriate electronic communication with assessors?
	- Emails should have a meaningful subject.
- [ ] Handed in the assignment before the due time (Friday tutorial 9am)?

### Active Participation (3%)

A small component of the course assessment will be for active participation in practicals and tutorials.
Since the majority of the course material is open source, issues filed against errors found in practicals, tutorials and lectures, and contributions to fixing those errors will count to this component.

### Major Project (15%)

In this course, the following next-generation sequencing (NGS) datasets/protocols will be in detail:

1. Whole genome sequencing/Resequencing
2. Transcriptome Sequencing (RNAseq)
3. DNA Methylation/Bisulfite Sequencing
4. Enrichment/Capture sequencing (Methyl-capture, ChIPseq, RIPseq)
5. Metagenomics/Microbial profiling

Each of these NGS approaches uses similar programs and analysis approaches, such as quality control (quality and sequencing adapter trimming), genome alignment, and downstream visualisation and statistical methods. For the major project, you will take a published dataset and complete all the analysis tasks (from raw data to final results) during the course.

You have the freedom to choose any dataset you would like before we start week 5's practical (NGS data practical), or choose from four datasets that we have chosen. You will be asked assessment questions on your data during practical assessments throughout the semester, and a final report will be due the week before the end of semester (Week 12)

## Useful Links

[How To Ask Questions The Smart Way](http://www.catb.org/esr/faqs/smart-questions.html)

[How to write a good bug report](https://musescore.org/en/developers-handbook/how-write-good-bug-report-step-step-instructions)
