---
title: "Assignment 5 [20 marks]"
author: "Biotech 7005: Bioinformatics & Systems Modelling"
date: "Due 5pm, 30 September 2019"
output: 
  html_document: 
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
											warning = FALSE, message = FALSE)
```


```{r}
library(Gviz)
library(tidyverse)
library(pander)
```


## Instructions [1 mark]

Submit your answers as a single `rmarkdown` file. Please ensure that your student number appears in the file name.


## Question 1 [6 marks]

```{r, fig.cap="*Transcripts for question 1, with different transcription start sites but the same termination site. The direction of transcription is shown by the arrows*", fig.height=1.5}
trans <- data.frame(
chromosome = "chrX",
start = c(100, 900, 1500, 1500),
end = c(1000, 1000, 1800, 1800),
width = c(900, 100, 300, 300),
strand = "+",
gene = "Q1",
transcript = c("Transcript A", "Transcript B")[c(1, 2, 1,2)],
genome = "hg19"
)
gTrack <- GeneRegionTrack(trans, name = "Q1")
plotTracks(gTrack, 
					 arrowHeadWidth = 50,
					 transcriptAnnotation = "transcript", 
					 cex.title = 1, 
					 cex = 4,
					 fill = rgb(0, 0.5, 0.8))
```


Our gene of interest has two possible transcripts, A and B.
The shorter transcript (B) is 400bp long, whilst the longer transcript (A) is 1200bp long.
In our RNA Seq experiment using 150bp single-end reads, we observed 100 reads in each of our two conditions, however:

- in our Control sample **the reads all came from transcript B.**
- in our Treated sample **the reads all came from transcript A.**

a. Would you consider this gene to be differentially expressed? Explain your reasoning.
b. Imagine these same samples were hybridised to Affymetrix 3' Microarrays, and note that these transcripts terminate in the same 3' exons. Would you expect to see differential expression in this microarray experiment? Why, or why not?
c. Describe a pair of hypothetical transcripts where a 3' microarray would be able to detect differential transcript usage?

## Question 2 [7 marks]

Ensure you include your `R` code for any calculations in this question.

```{r}
df <- data_frame(Replicate = paste("Replicate", rep(1:3, each = 2)),
								 Group = rep(c("Treated", "Control"), times = 3),
								 `$C_T$ FOXO1` = c(21.2, 19.7, 19.3, 17.3, 22.5, 19.9),
								 `$C_T$ ACTB` = c(7.1, 7.3, 6.6, 6.7, 8.7, 8.7),
								 `$\\Delta C_T$` = "")
```


For the following table, our mRNA of interest is from FOXO1, and we have used ACTB as the housekeeper.

a. Calculate the $\Delta C_T$ value for each replicate and treatment. 


```{r}
pander(df, style = "rmarkdown")
```


b. Was FOXO1 or ACTB more abundant in the original samples? Explain in reference to $C_T$ values.
c. Now calculate the $\Delta \Delta C_T$ values for each replicate 

| Replicate | $\Delta \Delta C_T$ |
| --------- | ------------------- |
| Replicate 1 |                   |
| Replicate 2 |                   |
| Replicate 3 |                   |

d. Conduct a $T$-test on the $\Delta \Delta C_T$ values using `R`
e. Referring to the above results, do you think FOXO1 was up or down regulated in the treated samples? 

## Question 3 [6 marks]

Discuss 3 advantages that RNA-Seq data has over microarray data.
